Allow customizing fetchurl hashedMirrors

jherland opened the pull request:
Nix users behind a firewall may have difficulties reaching the default mirrors referenced from `pkgs/build-support/fetchurl/mirrors.nix`. Additionally, Nix users inside corporations may need to prefer company-internal mirrors.

This PR exposes the set of mirrors used by `fetchurl` via the nixpkgs config, allowing users to customize it via their `~/.config/nixpkgs/config.nix`. The default mirror list is made available in `config.mirrors`, and users may remove/add/replace/adjust the mirrors via the module system, as shown in the documented example.

@infinisil edit: This PR is now changed to only allow customisation of `hashedMirrors`

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [x] Added a release notes entry about the new `config.mirrors` attribute
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jherland commented:
> How would you feel about just dropping the mirrors that have caused you issues?

Although dropping problematic mirrors would help some cases, allowing customizations to the set of mirrors remains a primary concern for us: We're working in a company scenario where access to external urls is deliberately blocked by firewalls, and package sources must be fetched through company-internal mirrors.

The customizations we want to apply to the mirror list include e.g. replacing `hashedMirrors = ["https://tarballs.nixos.org"]` with `hashedMirrors = [COMPANY_INTERNAL_MIRROR]`. In short, changes that would not be appropriate to merge into nixpkgs.

Adding this customization point allows us to get away with a configuration change rather than having to patch nixpkgs in our deployment.

> Then when all of these entries have only one URL we can think about dropping the mechanism entirely.

Do you mean dropping the `mirror://` scheme altogether? Even if each mirror only has one URL associated with it, I still appreciate the abstraction it offers, especially for mirrors that are used by a _lot_ of packages.

emilazy commented:
I’ve put up https://github.com/NixOS/nixpkgs/pull/409110.

> Although dropping problematic mirrors would help some cases, allowing customizations to the set of mirrors remains a primary concern for us: We're working in a company scenario where access to external urls is deliberately blocked by firewalls, and package sources must be fetched through company-internal mirrors.
> 
> The customizations we want to apply to the mirror list include e.g. replacing `hashedMirrors = ["https://tarballs.nixos.org"]` with `hashedMirrors = [COMPANY_INTERNAL_MIRROR]`. In short, changes that would not be appropriate to merge into nixpkgs.

Right. But the vast majority of packages don’t use this mechanism at all; we’re increasingly preferring to fetch sources from GitHub and other source repositories over release tarballs, for instance. So you need a solution that works for packages that don’t use `mirror://` anyway.

`hashedMirrors` is a different kind of thing IMO; a customization point for that would make sense to me. For the rest, I’m not sure I see the value: you can’t avoid downloading external URLs even if you override all of these mirrors. I think the best solution there is to set up a selective caching HTTPS proxy, which is something Nix/Nixpkgs should support quite well at this point.

The remaining value I see in `mirror://` is in templating mirrors that are used across the tree and that either might change or need fallbacks for things that get moved to an archive, yeah. But I don’t think there’s all too much of that; it’s hard to imagine ftp.gnu.org will ever move, and if it did we could just do a tree‐wide find and replace. The intersection of “fairly widely used throughout the tree” and “might require a fallback archive URL” is pretty much Linux distros, and even for Debian that’s only some ~40 uses throughout the tree, and they’d probably be happier with a `fetchurl` wrapper handling it anyway given the fairly baroque URL format. So honestly I don’t think there’s much reason to keep it around.

oxij commented:
To quote https://github.com/NixOS/nixpkgs/pull/409110#issuecomment-2896703942 which is relevant here:

> Also, philosophically, I disagree with [#409010 (comment)](https://github.com/NixOS/nixpkgs/pull/409010#issuecomment-2894259994) and [#409010 (comment)](https://github.com/NixOS/nixpkgs/pull/409010#issuecomment-2895051455). The fact that CDNs are common in 2025 is not an argument for making Nixpkgs less resilient to censorship and networking issues. IMHO, with CDNs and GitHub centralization we should be adding more mirrors and implement the ability to specify alternative non-GitHub `src`s for `fetchFromGitHub`, when such things are available. Which, for many projects, they are.
> 
> The fact that several companies can single-handedly prevent you from updating your machines does not seem that nice of a thing to me.

balsoft commented:
> hashedMirrors is a different kind of thing IMO; a customization point for that would make sense to me. For the rest, I’m not sure I see the value: you can’t avoid downloading external URLs even if you override all of these mirrors. I think the best solution there is to set up a selective caching HTTPS proxy, which is something Nix/Nixpkgs should support quite well at this point.

It's currently handled by the same mechanism, and adding an override specifically for it is _more work_ than adding an override for the entire mirror list, all for less functionality. If in the future we want to get rid of non-hashed mirrors entirely, this PR doesn't make it any more difficult to do.

philiptaron commented:
I'm loathe to merge this since it's my belief that https://github.com/NixOS/nixpkgs/pull/410186 allows your usecases to complete without this complication of the mirror system.

balsoft commented:
@philiptaron 

> I'm loathe to merge this since it's my belief that https://github.com/NixOS/nixpkgs/pull/410186 allows your usecases to complete without this complication of the mirror system.

Actually, it's not possible to use `rewriteURL` for `hashedMirrors`; the only way to override them right now is with an impure environment variable, which doesn't play well with multi-user Nix and is in general quite hacky. I still think this PR should be merged, as it allows for customization of `hashedMirrors`; even if we drop non-hashed mirrors in the future, this would still be useful for multi-user Nix setups (or otherwise to avoid setting the env variable).

infinisil reviewed (changes requested):
> It's currently handled by the same mechanism, and adding an override specifically for it is more work than adding an override for the entire mirror list, all for less functionality. If in the future we want to get rid of non-hashed mirrors entirely, this PR doesn't make it any more difficult to do.

I have to disagree with that:
- Making only `hashedMirrors` overridable is not substantially harder than making all `mirrors` overridable, see suggestion below
- The goal should not be to have the most functionality, but to have functionality according to what we need. And @jherland [confirmed](https://github.com/NixOS/nixpkgs/pull/409010#issuecomment-2894481840) that they only need `hashedMirrors` to be overridable:
  > The customizations we want to apply to the mirror list include e.g. replacing `hashedMirrors = ["https://tarballs.nixos.org"]` with `hashedMirrors = [COMPANY_INTERNAL_MIRROR]`.
- This PR in its current state _does_ make it harder to get rid of non-hashed mirrors in the future, because it introduces a public interface for them, which comes with expectation of stability and deprecation periods, which does require maintenance and consideration to handle properly.

So, I think we should just expose an override interface for `hashedMirrors`, which is also a great step towards avoiding conflation with `mirrors`. The fact that those are specified with the same attribute set is really just an implementation detail anyways, it's not fit for an interface.

If there is also a requirement to override non-hashed `mirrors` (now or in the future), a separate PR would be more appropriate to not conflate the discussions of the two any more.

PS: I'd like to refer to [C4](https://rfc.zeromq.org/spec/42/), which I hope to see used more, point 2.3.2:
> A patch SHOULD be a minimal and accurate answer to exactly one identified and agreed problem.

infinisil committed:
stdenv: Pass rewriteURL to freebsd and native stdenvs

Was missing from abda866f172173b6fd16ccec9b15fb1cbb649cb2

While the native stdenv's eval seems to be broken anyways,
the freebsd one at least evaluates again with this

infinisil committed:
fetchurl: Allow hashed mirror overriding with config.hashedMirrors

Allows having alternate hashed mirrors as fallbacks. Useful in case the
default hashed mirror is not accessible or doesn't have everything
needed.

Co-authored-by: Johan Herland <johan.herland@tweag.io>
Co-authored-by: Yuriy Taraday <yuriy.taraday@tweag.io>
Co-authored-by: Alexander Bantyev <balsoft@balsoft.ru>
